### PR TITLE
Add benchmark for IndexNameExpressionResolver

### PR DESCRIPTION
**Pull request by gmarouli**
Recently we caused a performance regression with https://github.com/elastic/elasticsearch/commit/4c15cc077887d00ecf0e02c39b42cf01874ab6c4. 

This commit has been reverted and this commit https://github.com/elastic/elasticsearch/commit/d5265bef572eaa87cc07b861ad00c74f8a955fbf has improved its performance.

In this PR we introduce a small microbenchmark that we hope will shorten the feedback loop during development.

Example of run over 3 different commits on a local machine with a cluster state with 10000 data streams with 1000 backing indices each, and 1000 standalone indices:

```
Initial state: 057062bcae2                                                                  (resourceMix)  Mode  Cnt      Score      Error  Units
IndexNameExpressionResolverBenchmark.resolveAllStarToConcreteIndices        10000|       1000  avgt   15  10394.024 ± 685.879  ms/op
IndexNameExpressionResolverBenchmark.resolveResourcesListToConcreteIndices  10000|       1000  avgt   15   9370.274 ± 491.076  ms/op

Added wrapper: 4c15cc07788                                                             (resourceMix)  Mode  Cnt      Score      Error  Units
IndexNameExpressionResolverBenchmark.resolveAllStarToConcreteIndices        10000|       1000  avgt   15  10323.434 ± 555.209  ms/op
IndexNameExpressionResolverBenchmark.resolveResourcesListToConcreteIndices  10000|       1000  avgt   15  10836.716 ± 433.244  ms/op

Current: feea0a09b87                                                                      (resourceMix)  Mode  Cnt      Score      Error  Units
IndexNameExpressionResolverBenchmark.resolveAllStarToConcreteIndices        10000|       1000  avgt   15  11130.473 ± 504.111  ms/op
IndexNameExpressionResolverBenchmark.resolveResourcesListToConcreteIndices  10000|       1000  avgt   15   7547.232 ± 530.436  ms/op
```

### CODE REVIEW

File: benchmarks/src/main/java/org/elasticsearch/benchmark/indices/resolution/IndexNameExpressionResolverBenchmark.java [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+        return indexMetadata;
+    }
+
+    private IndexNameExpressionResolver resolver;
+    private ClusterState clusterState;
+    private String[] indices;
+
+    @Benchmark
+    public String[] resolveResourcesListToConcreteIndices() {
+        return resolver.concreteIndexNames(clusterState, new IndicesRequest() {
```
Comment (original-brownbear) [inline benchmarks/src/main/java/org/elasticsearch/benchmark/indices/resolution/IndexNameExpressionResolverBenchmark.java:106]: This is a bit of a technicality but lets make this more realistic and make the `IndicesRequest` implementation here a `static` named class and have it copy the indices list in it in the constructor? :)

**Review by original-brownbear — COMMENTED** [review]
This looks quite cool thanks LGTM.
Maybe do the static inner class thing if you have a sec :) But this is a great start to add problematic scenarios to as we encounter them!

**Comment by gmarouli** [comment]
@elasticmachine update branch

**Comment by gmarouli** [comment]
Added one more benchmark to test a mixed request with a looooot of concrete expressions but also a wildcard. The reason to test this is that we have some optimised paths for long list of concrete indices compared to mixed expressions.

**Comment by gmarouli** [comment]
@elasticmachine update branch

**Review by original-brownbear — APPROVED** [review]
LGTM this looks quite nice thanks :)
It's quite interesting to see the results too, these are shockingly bad aren't they? :) We should look into that. `resolveResourcesListToConcreteIndices` seems like it represents when currently happens in the wild with security in the mix?

**Pull request merged**